skimage/segmentation/_watershed_cy.pyx

```diff
@@ -117,13 +122,17 @@ def watershed_raveled(cnp.float64_t[::1] image,
     cdef Py_ssize_t index = 0
     cdef Py_ssize_t neighbor_index = 0
     cdef DTYPE_BOOL_t compact = (compactness > 0)
+    cdef cnp.float64_t dbl_min = np.finfo(np.float64).min
 
     cdef Heap *hp = <Heap *> heap_from_numpy2()
 
     with nogil:
         for i in range(marker_locations.shape[0]):
             index = marker_locations[i]
-            elem.value = image[index]
+            if from_minima:
+                elem.value = image[index]
+            else:
+                elem.value = dbl_min
```

Comment:
@lagru, thanks for the review. There was indeed a bug.
#6632 and #4240 were working on commit `6ef145b`, when I fixed the bug pointed out by @jni, they stopped working because of an error.

When I fixed the error, only #6632 worked, which confused me because #4240 was supposed to be fixed too. So I noticed the `_diff_neighbors` was checking the neighbors' labels but not itself, so I fixed it and got #4240 and #6279 working too :)

> You set all markers to the minimal possible double value so that when the actual watershed algorithm starts drawing from the heap, new pixels will always be sorted after the initial markers, regardless of their pixel value?

Yes.

> In that case, I am really curious about the intended effect here. At a first glance it seems un-intuitive to me why markers are a special case whose neighborhood must be evaluated first. 

The watersheding of a marker $p$ will only label pixels with values greater than their own, if you don't set the $p$'s value to $-\infty$ it won't label a neighboring pixel $q$ with lower-value when there's a competing marker $t$ with a path to $q$ lower than $p$'s value. The intended behavior is that the water from the markers should drip down the bottom of neighboring basins (neighboring pixels with lower values).

For reference, [[1]](https://www.researchgate.net/profile/Serge-Beucher/publication/230837870_The_Morphological_Approach_to_Segmentation_The_Watershed_Transformation/links/00b7d5319b26f3ffa2000000/The-Morphological-Approach-to-Segmentation-The-Watershed-Transformation.pdf), Section IV A 4, $$W_{-1}(g) = M$$ The markers are set to level -1 (they consider the image has non-negative values).
This is shown in Figure 16, where the `modified function` (hatched region) is zero on the markers (selected minima).

I changed the compact watershed test again and added a "barrier" with a greater value than the markers' regions, so the result is not an implementation detail as before (tie-zones). I explain it in the test documentation.

Reply:
Thanks for the detailed response. I haven't finished reviewing everything yet but I can confirm that https://github.com/scikit-image/scikit-image/issues/6632 and https://github.com/scikit-image/scikit-image/issues/4240#issuecomment-1235358536 return the same result regardless of the marker location or the parameter `watershed_line`. Though the output is changed.

Would you be willing to add new tests for both cases to avoid this kind of bug in the future? :pray: E.g. for the `watershed_line` a test that asserts that the labels returned with `watershed_line=True` are a strict subset of the labels returned with `watershed_line=False`? I am happy to do so as well if you are so busy right now!

Reply:
Hi @lagru, I added the tests. I decreased the offset of the markers from 6 to 4 for the #6632 test.
With 6, it had a 3 pixels difference, it's not much, but I wanted to avoid a magic number in the tests.

It's a tricky example because the difference between the radius and the distance between their center changes the boundary of the results.

My understanding is the following, for two adjacent but not overlapping circles, the markers can have an offset of, at most, the difference between their radii. If it's greater than the radii difference, the neighbors around the marker of the larger circle won't let it propagate before the smaller circle. Note that, this problem is specific to when using EDT for the watershed, and when there's overlap it gets much trickier because it depends on how the EDT changes with the overlap.

Reply:
I updated this test in https://github.com/scikit-image/scikit-image/pull/7071/commits/1e3b8ed2a156603b6619b293e5fb2b87521e44c0 to use the actual bug scenario with an y-offset of 6. Instead of comparing to the 0-offset case, I'm checking the result with regionproperty measures eccentricity and num_pixels.